### ISSUE TITLE
conv4 error on machine with cuda

### ISSUE THREAD

**Issue by IanButterworth** [ianshmean/YOLO.jl#9]
@Ybakman it seems like the second argument to `conv4` needs to be matched to the first, which doesn't happen on a machine with CUDA (first time I'm trying it)

Hopefully a simple fix, but I thought I'd check with you (or @denizyuret) 

The offending line:
https://github.com/ianshmean/YOLO.jl/blob/f1944fda4d1518b9442ff3bd607605128c4b49b9/src/models/v2_tiny/v2_tiny.jl#L48-L49

```
Loading and running YOLOv2_tiny_voc pretrained model: Error During Test at /home/ian/Documents/GitHub/YOLO.jl/test/runtests.jl:30
  Got exception outside of a @test
  MethodError: no method matching conv4(::Knet.KnetArray{Float32,4}, ::Array{Float32,4}; stride=1, padding=1)
  Closest candidates are:
    conv4(::Knet.KnetArray{T,N} where N, !Matched::Knet.KnetArray{T,N} where N; handle, alpha, o...) where T at /home/ian/.julia/packages/Knet/ztjo0/src/conv.jl:37
    conv4(::##1054, !Matched::AutoGrad.Value{##1055}; o...) where {##1054, ##1055} at none:0
    conv4(!Matched::AutoGrad.Value{##1054}, ::##1055; o...) where {##1054, ##1055} at none:0
    ...
  Stacktrace:
   [1] (::YOLO.v2_tiny.Conv)(::Array{Float32,4}) at /home/ian/Documents/GitHub/YOLO.jl/src/models/v2_tiny/v2_tiny.jl:48
   [2] (::YOLO.v2_tiny.Chain)(::Array{Float32,4}) at /home/ian/Documents/GitHub/YOLO.jl/src/models/v2_tiny/v2_tiny.jl:45
   [3] top-level scope at /home/ian/Documents/GitHub/YOLO.jl/test/runtests.jl:85
   [4] top-level scope at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.2/Test/src/Test.jl:1113
```

**Comment by Ybakman** [comment]
I suppose the input given the model is a Julia array. To run the model on GPU,  the input also should be a KnetArray. You can convert the array into KnetArray by simply KnetArray{Float32}(x). x is Array{Float32} Input. Also, you can check if the machine has GPU by Knet.gpu() function. If it is >= 0, GPU exists.

**Comment by Ybakman** [comment]
Also, If you use minibatch you can specify the type of array by xtype and ytype keyword arguments.
For instance, minibatch(input,label,MINIBATCH_SIZE; xtype = KnetArray{Flaot32}). Now, input is a KnetArray and label is julia array

**Comment by Ybakman** [comment]
In the loadweights! function, I converted the conv layer array to KnetArray, when gpu exists.

**Comment by IanButterworth** [comment]
Makes sense! I just implemented the `KnetArray` switch for `LoadedDataset.imstack_mat` over in https://github.com/ianshmean/YOLO.jl/pull/10

The default test tries to push 10 images through at once, so perhaps we should be `minibatch`-ing as you suggest

**Comment by Ybakman** [comment]
Definitely, if we don't use it, Gpu memory is not enough to be loaded the whole dataset. Minibatch function simply carries input to Gpu and remove previous data from Gpu.

**Comment by Ybakman** [comment]
Also, I want to emphasize if Knet.gpu() == 0, the machine has gpu. When gpu doesn't exist , Knet.gpu() == -1.

**Comment by IanButterworth** [comment]
Ah, I missed that.. Thanks for highlighting it.

This issue is now fixed with the merging of #10, but we do need to incorporate `minibatch`-ing properly.
